sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureFileSystem.java

```diff
@@ -0,0 +1,259 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.blob.nio;
+
+import com.azure.core.http.HttpClient;
+import com.azure.core.http.policy.HttpLogDetailLevel;
+import com.azure.core.http.policy.HttpLogOptions;
+import com.azure.core.util.CoreUtils;
+import com.azure.storage.blob.BlobServiceClient;
+import com.azure.storage.blob.BlobServiceClientBuilder;
+import com.azure.storage.common.StorageSharedKeyCredential;
+import com.azure.storage.common.policy.RequestRetryOptions;
+import com.azure.storage.common.policy.RetryPolicyType;
+
+import java.io.IOException;
+import java.nio.file.FileStore;
+import java.nio.file.FileSystem;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.nio.file.WatchService;
+import java.nio.file.attribute.UserPrincipalLookupService;
+import java.nio.file.spi.FileSystemProvider;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * In the hierarchy of this file system, an {@code AzureFileSystem} corresponds to an Azure Blob Storage account. A
+ * file store is represented by a container in the storage account. Each container has one root directory.
+ *
+ * Closing the file system will not block on outstanding operations. Any operations in progress will be allowed to
+ * terminate naturally after the file system is closed, though no further operations may be started after the parent
+ * file system is closed.
+ * {@inheritDoc}
+ */
+public final class AzureFileSystem extends FileSystem {
+    // Configuration constants for blob clients.
+    public static final String AZURE_STORAGE_ACCOUNT_KEY = "AzureStorageAccountKey";
+    public static final String AZURE_STORAGE_SAS_TOKEN = "AzureStorageSasToken";
+    public static final String AZURE_STORAGE_HTTP_LOG_DETAIL_LEVEL = "AzureStorageHttpLogDetailLevel";
+    public static final String AZURE_STORAGE_MAX_TRIES = "AzureStorageMaxTries";
+    public static final String AZURE_STORAGE_TRY_TIMEOUT = "AzureStorageTryTimeout";
+    public static final String AZURE_STORAGE_RETRY_DELAY_IN_MS = "AzureStorageRetryDelayInMs";
+    public static final String AZURE_STORAGE_MAX_RETRY_DELAY_IN_MS = "AzureStorageMaxRetryDelayInMs";
+    public static final String AZURE_STORAGE_RETRY_POLICY_TYPE = "AzureStorageRetryPolicyType";
+    public static final String AZURE_STORAGE_SECONDARY_HOST = "AzureStorageSecondaryHost";
+    public static final String AZURE_STORAGE_BLOCK_SIZE = "AzureStorageBlockSize";
+    public static final String AZURE_STORAGE_DOWNLOAD_RESUME_RETRIES = "AzureStorageDownloadResumeRetries";
+    public static final String AZURE_STORAGE_USE_HTTPS = "AzureStorageUseHttps";
+    public static final String AZURE_STORAGE_HTTP_CLIENT = "AzureStorageHttpClient"; // undocumented; for test.
```

Comment:
Do we want to leave this undocumented? I think allowing a customer to perform as many options as when building a general client is best.

Reply:
I don't feel strongly about leaving it undocumented. The reason I didn't add more full support for it was because I wasn't sure how deep to go into configuring the client for them and I just wanted something for testing more immediately. Should we just accept parameters for the proxy configs and the kind of client they want? Or should we just accept a whole preconfigured client? The former requires lots of options. The latter requires more knowledge from the customer on the underlying client which isn't good.

Reply:
For the initial preview I think leaving this simpler, the option where a pre-configured `HttpClient` is the only option available, would be the better option. The latter would require a lot of knowledge of the configuration options available and may cover a smaller group than allowing an overarching pre-configured client.